eventstore/dynamodb/native/src/main/java/org/occurrent/eventstore/dynamodb/nativedriver/DynamoDBEventStore.java

```diff
@@ -0,0 +1,262 @@
+                  .build())
+          .attributeDefinitions(
+              AttributeDefinition.builder().attributeName(ATTRIBUTE_STREAM)
+                  .attributeType(ScalarAttributeType.S).build(),
+              AttributeDefinition.builder().attributeName(ATTRIBUTE_VERSION)
+                  .attributeType(ScalarAttributeType.N).build())
+          .provisionedThroughput(ProvisionedThroughput.builder().readCapacityUnits(readCapacity)
+              .writeCapacityUnits(writeCapacity).build())
+          .build());
+    } catch (ResourceInUseException e) {
+      // ignore existing table
+    }
+    return new DynamoDBEventStore(dynamoDB, tableName, timeRepresentation);
+  }
+
+  public void deleteTable() {
+    try {
+      dynamoDB.deleteTable(DeleteTableRequest.builder().tableName(tableName).build());
+    } catch (ResourceNotFoundException e) {
+      // ignore
+    }
+  }
+
+  @Override
+  public WriteResult write(String streamId, Stream<CloudEvent> events) {
+    return write(streamId, WriteCondition.anyStreamVersion(), events);
+  }
+
+  @Override
+  public WriteResult write(String streamId, WriteCondition writeCondition,
+      Stream<CloudEvent> events) {
+    QueryResponse res = dynamoDB.query(QueryRequest.builder()
+        .tableName(tableName)
+        .attributesToGet(ATTRIBUTE_VERSION)
+        .scanIndexForward(false)
+        .limit(1)
+        .keyConditions(Map.of(ATTRIBUTE_STREAM, Condition.builder()
+            .attributeValueList(AttributeValue.builder().s(streamId).build())
+            .comparisonOperator(ComparisonOperator.EQ)
+            .build()))
+        .build());
+    AtomicLong version = res.items().isEmpty()
+        ? new AtomicLong(0)
+        : new AtomicLong(Long.parseLong(res.items().get(0).get(ATTRIBUTE_VERSION).n()));
+    long oldStreamVersion = version.intValue();
+
+    if (writeCondition instanceof StreamVersionWriteCondition) {
+      org.occurrent.condition.Condition<Long> condition = ((StreamVersionWriteCondition) writeCondition).condition;
+      if (condition != null && !LongConditionEvaluator.evaluate(condition, oldStreamVersion)) {
+        throw new WriteConditionNotFulfilledException(streamId, oldStreamVersion, writeCondition, String.format("%s was not fulfilled. Expected version %s but was %s.", WriteCondition.class.getSimpleName(), writeCondition, version.longValue()));
+      }
+    }
+    List<Pair<String,Long>> eventData = events
+        .map(e -> {
+          long v = version.incrementAndGet();
+          return new Pair<>(OccurrentCloudEventMongoDocumentMapper.convertToDocument(timeRepresentation, streamId, v, e).toJson(), v);
+        })
+        .collect(Collectors.toList());
+
+    // TODO: optimize
+    if (eventData.isEmpty()) {
+      return new WriteResult(streamId, oldStreamVersion, oldStreamVersion);
+    }
+
+    try {
+      dynamoDB.transactWriteItems(TransactWriteItemsRequest.builder()
```

Comment:
Yes I'm aware of this and this will most likely be a configuration option how to write. Ideally I would like adding events to an eventstream to be an atomic operation and initially I had one row per transaction (ie several events), but then I had to change this in order to conform to what seems to be the rule in occurrent that every single event should increment the version by one.

@johanhaleby Related to this, I was  confused by this: EventStream<CloudEvent> read(String streamId, int skip, int limit);

Should skip be the number of events to skip or should this be the version number? If its the version number, shouldnt this be a long? Does the version number have to equal the number of events?

When doing eventsourcing I usually consider all the events generated by a command to be an atomic update of the eventstream, ie having versions between the start and the end of the transaction does not necessarily make sense.

Reply:
>I had one row per transaction (ie several events), but then I had to change this in order to conform to what seems to be the rule in occurrent that every single event should increment the version by one.

Yeah the problem/tradeoff is that the minute you try to fulfil the transactional correctness requirement, you run into a set of problems:
- TransactWriteItems doubles the charges for everything, which is a massive loss
- You need something useful to make the write contingent on (i.e. the expectedversion etc - if you instead are checking that the previous item is present and the one you are writing is not, even the basic coding gets complex)

In Equinox, the schema resolves the forces by having the notion of a _Tip_ per stream, which gates all writes going through:
- one could keep an event counter in it, but I have an etag string (this allows one to rolling transactionally correct updates without having to write a new event every time). Where you are writing event 0, the condition is that the Tip does not exist
- if you are persisting multiple events in one write, they can all get appended in a single Put/Update call

In addition to working for larger cases, it has the following key properties for normal use:
- small streams are a single item that can be loaded via a single GetItem roundtrip
- TransactWriteItems only becomes required when the tip overflows
- minimum storage overhead

One thing it does complicate is the fact that the DDB Streams output will emit a full copy of the Tip for every update, e.g. if you are adding 2 events to a Tip that has one event already, the DDB streams output will be a DDB Streas event with the full Item (which hosts 3 events, but only 2 are new)

The other thing to bear in mind is that having >1 event per item means you need a good story about when you are writing 201K of events on top of 200K of existing events.

I would caution against having a mode switch in your implementation; testing, reasoning and talking about the code becomes a nightmare. Better to have a single impl that can deal with your use cases efficiently and test, tune and validate that. (The other reason I say that is that I fundamentally believe that an event per Item schema is just worse than useless in terms of cost and efficiency too)

>Should skip be the number of events to skip or should this be the version number? If its the version number, shouldnt this be a long? Does the version number have to equal the number of events?

I use longs for event indexes; ESDB etc does too. In practice the CUs and latency it costs to read more than 2m events make it irrelevant (and there are fixed limits to how much can be held in a logical partition (10GB is it?), so any design that is predicated on unlimited stream lengths is not even theoretically implementable)